src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -3781,7 +3781,11 @@ void MergeTreeData::delayInsertOrThrowIfNeeded(Poco::Event * until, ContextPtr q
         max_k = settings->inactive_parts_to_throw_insert - settings->inactive_parts_to_delay_insert;
         k = k_inactive + 1;
     }
-    const UInt64 delay_milliseconds = static_cast<UInt64>(::pow(settings->max_delay_to_insert * 1000, static_cast<double>(k) / max_k));
+
+    const UInt64 max_delay_milliseconds = (settings->max_delay_to_insert > 0 ? settings->max_delay_to_insert * 1000 : 1000);
+    /// max() as a save guard here
+    const UInt64 delay_milliseconds
+        = std::max(max_delay_milliseconds, static_cast<UInt64>(::pow(max_delay_milliseconds, static_cast<double>(k) / max_k)));
```

Comment:
I cannot understand how this possibly can fix something.
The visible error in logs I saw:
```
Delaying inserting block by 9223372036854775808 ms. because there are 199 parts and their average size is 1.85 GiB
```
Why max here?

FYI: initially I don't understand how can we use time argument in pow function. It does not have any sense. Like, what's (millisecond) ^ (1/3) ?
For me, something like `max_delay_to_insert * pow(some_constant, (k / max_k) - 1))` should work (where `some_constant > 1, k >= max_k`. So that `k = 0 => result is max_delay_to_insert / some_constant; k = max_k => result is max_delay_to_insert`. It may be logical to have also setting like `min_delay_to_insert` and set `some_constant := max_delay_to_insert / min_delay_to_insert`

Reply:
Code above with `k`, `max_k` and `k_inactive` looks like some trash :(

Reply:
Even more clean formula:
`delay = min_delay_to_insert * pow(max_delay_to_insert / min_delay_to_insert, 1.0 - k / max_k)`

Reply:
> I cannot understand how this possibly can fix something. The visible error in logs I saw:
> 
> ```
> Delaying inserting block by 9223372036854775808 ms. because there are 199 parts and their average size is 1.85 GiB
> ```
> 
> Why max here?
> 
> FYI: initially I don't understand how can we use time argument in pow function. It does not have any sense. Like, what's (millisecond) ^ (1/3) ? For me, something like `max_delay_to_insert * pow(some_constant, (k / max_k) - 1))` should work (where `some_constant > 1, k >= max_k`. So that `k = 0 => result is max_delay_to_insert / some_constant; k = max_k => result is max_delay_to_insert`. It may be logical to have also setting like `min_delay_to_insert` and set `some_constant := max_delay_to_insert / min_delay_to_insert`

(1) Yes, it's stupid, - of course, it should be min
(2) Because, this formula is not clear and depends on number of `active' or 'outdated' parts, - I wanted to make it clean in other PR. Here it's just a fast fix.

Reply:
ok